python/lsst/daf/butler/registry/_sqlRegistry.py

```diff
@@ -0,0 +1,1849 @@
+                               for dataId in progress.wrap(dataIds,
+                                                           f"Expanding {storage.datasetType.name} data IDs")]
+        else:
+            expandedDataIds = [DataCoordinate.standardize(dataId, graph=storage.datasetType.dimensions)
+                               for dataId in dataIds]
+        try:
+            refs = list(storage.insert(runRecord, expandedDataIds))
+        except sqlalchemy.exc.IntegrityError as err:
+            raise ConflictingDefinitionError(f"A database constraint failure was triggered by inserting "
+                                             f"one or more datasets of type {storage.datasetType} into "
+                                             f"collection '{run}'. "
+                                             f"This probably means a dataset with the same data ID "
+                                             f"and dataset type already exists, but it may also mean a "
+                                             f"dimension row is missing.") from err
+        return refs
+
+    def getDataset(self, id: int) -> Optional[DatasetRef]:
+        """Retrieve a Dataset entry.
+
+        Parameters
+        ----------
+        id : `int`
+            The unique identifier for the dataset.
+
+        Returns
+        -------
+        ref : `DatasetRef` or `None`
+            A ref to the Dataset, or `None` if no matching Dataset
+            was found.
+        """
+        ref = self._managers.datasets.getDatasetRef(id)
+        if ref is None:
+            return None
+        return ref
+
+    @transactional
+    def removeDatasets(self, refs: Iterable[DatasetRef]) -> None:
+        """Remove datasets from the Registry.
+
+        The datasets will be removed unconditionally from all collections, and
+        any `Quantum` that consumed this dataset will instead be marked with
+        having a NULL input.  `Datastore` records will *not* be deleted; the
+        caller is responsible for ensuring that the dataset has already been
+        removed from all Datastores.
+
+        Parameters
+        ----------
+        refs : `Iterable` of `DatasetRef`
+            References to the datasets to be removed.  Must include a valid
+            ``id`` attribute, and should be considered invalidated upon return.
+
+        Raises
+        ------
+        AmbiguousDatasetError
+            Raised if any ``ref.id`` is `None`.
+        OrphanedRecordError
+            Raised if any dataset is still present in any `Datastore`.
+        """
+        progress = Progress("lsst.daf.butler.Registry.removeDatasets", level=logging.DEBUG)
+        for datasetType, refsForType in progress.iter_item_chunks(DatasetRef.groupByType(refs).items(),
+                                                                  desc="Removing datasets by type"):
+            storage = self._managers.datasets.find(datasetType.name)
+            assert storage is not None
```

Comment:
Using `self._managers.datasets[datasetType.name]` should remove the need for `assert`, there are few cases like this below.

Reply:
I didn't write any of this....